moja-global/FLINT.Cloud
# Implement pre-processing script for GCBM simulation

## Description

1. The PR implements the pre-processing script for GCBM demo run. Previously, the user had to upload hardcoded JSONs from `GCBM_Demo_Run` folder, we have eliminated this step, along with configuring the needed JSON values using rasterio. 2. The `config` endpoint lets an user change schema for the `gcbm_input.db`.

Steps to follow to test the changes:

1. avigate to the `local/rest_api_gcbm`

2. Run: `docker build --build-arg BUILD_TYPE=RELEASE --build-arg NUM_CPU=4 -t gcbm-api .`

3. Run: `docker run --rm -p 8080:8080 gcbm-api`

4. On a new terminal window, cd into `GCBM_New_Demo_Run`:
` cd path-to-unzipped-file`

5.  To create a new run: `curl -d "title=run4" -X POST http://localhost:8080/gcbm/new`
The title of a new simulation can be passed or the default title `simulation` will be used. (i.e the default title of the simulation is `simulation`, here the title `run4` is used)

6. Upload the files to the docker container :
`curl -F disturbances='@disturbances/disturbances_2011_moja.tiff' \
            -F disturbances='@disturbances/disturbances_2012_moja.tiff' \
            -F disturbances='@disturbances/disturbances_2013_moja.tiff' \
            -F disturbances='@disturbances/disturbances_2014_moja.tiff' \
            -F disturbances='@disturbances/disturbances_2015_moja.tiff' \
            -F disturbances='@disturbances/disturbances_2016_moja.tiff' \
            -F disturbances='@disturbances/disturbances_2018_moja.tiff' \
            -F classifiers='@classifiers/Classifier1_moja.tiff' \
            -F classifiers='@classifiers/Classifier2_moja.tiff' \
            -F db='@db/gcbm_input.db' \
            -F miscellaneous='@miscellaneous/initial_age_moja.tiff' \
            -F miscellaneous='@miscellaneous/mean_annual_temperature_moja.tiff' \
            -F title="run4" \
            http://localhost:8080/gcbm/upload`

7. Start the run: `curl -d "title=run4" -X POST http://localhost:8080/gcbm/dynamic`
Wait for 10-20 minutes for the run to finish, the end message should be something like: `insert SQLite`

8. To download the output zip folder, run: `curl -d "title=run4" -X POST http://localhost:8080/gcbm/download -L -o run4.zip` . Note you should change run4 to whatever you wish to name the simulation.

If you wish to change a table name in `gcbm_input.db`, we can do so by sending a post request to the `/config` endpoint.
```
`{
         "name" : "config",
         "request" : "POST",
         "header" : [],
         "body" : {
                      "mode" : "raw",
                      "JSON" : [
                             {
                                {
                                    "simulation_name": "run4",
                                           "tables" : {
                                             "classifier": {
                                                 "name": "play"
                                               }
                                         }
                                  }
                              }
                        ]

}
`
```
cc: @aornugent @Namyalg @ankitaS11 

Fixes #107 
## Type of change

Please delete options that are not relevant.


...

## Checklist:

- [x] I have tagged the reviewers in a comment below incase my pull request is ready for a review
- [x] I have signed the commit message to agree to Developer Certificate of Origin (DCO) (to certify that you wrote or otherwise have the right to submit your contribution to the project.) by adding "--signoff" to my git commit command.
